docs/static/transforming-data.asciidoc

```diff
@@ -0,0 +1,520 @@
+input {
+  tcp {
+    codec => fluent
+    port => 4000
+  }
+}
+--------------------------------------------------------------------------------
+
+
+//ALVIN: SHOULD WE ALSO COVER THE MSGPACK CODEC? I LOOKED AT THE DOCS FOR MSGPACK AND THERE'S NO USEFUL INFO IN THERE. 
+
+<<plugins-codecs-json,json codec>>::
+
+Decodes (via inputs) and encodes (via outputs) JSON formatted content, creating
+one event per element in a JSON array.
++
+The following config decodes the JSON formatted content in a file:
++
+[source,json]
+--------------------------------------------------------------------------------
+input {
+  file {
+    path => "/path/to/myfile.json"
+    codec =>"json"
+}
+--------------------------------------------------------------------------------
+
+
+//ALVIN: SHOULD WE INCLUDE JSON_LINES HERE, TOO, OR MAYBE MENTION IT?
+
+<<plugins-codecs-protobuf,protobuf codec>>::
+
+Reads protobuf encoded messages and converts them to Logstash events. Requires
+the protobuf definitions to be compiled as Ruby files.
++
+The following config decodes events from a Kafka stream:
++
+[source,json]
+--------------------------------------------------------------------------------
+kafka {
+  zk_connect => "127.0.0.1"
+  topic_id => "your_topic_goes_here"
+  codec => protobuf {
+    class_name => "Animal::Unicorn"
+    include_path => ['/path/to/protobuf/definitions/UnicornProtobuf.pb.rb']
+  }
+}
+--------------------------------------------------------------------------------
+
+
+//REVIEWERS: PLEASE DOUBLE-CHECK THE SYNTAX FOR THE PROTOBUF EXAMPLE. I CHANGED THE FORMATTING TO BE CONSISTENT WITH THE REST OF THE DOCUMENTATION HERE, BUT DIDN'T TEST THIS EXAMPLE.
+
+<<plugins-filters-xml,xml filter>>::
+
+Parses XML into fields.
++
+The following config parses the whole XML document stored in the `message` field:
++
+[source,json]
+--------------------------------------------------------------------------------
+filter {
+  xml {
+    source => "message"
+  }
+}
+--------------------------------------------------------------------------------
+
+
+[[field-extraction]]
+=== Extracting Fields and Wrangling Data
+
+The plugins described in this section are useful for extracting fields and
+parsing unstructured data into fields.
+
+Also see <<filter-plugins>> for the full list of available filter plugins. 
+
+<<plugins-filters-dissect,dissect filter>>::
+
+Extracts unstructured event data into fields by using delimiters. The dissect
+filter does not use regular expressions and is very fast. However, if the
+structure of the data varies from line to line, the grok filter is more
+suitable.
++
+The following config dissects messages that match the structure specified in
+the mapping:
++
+[source,json]
+--------------------------------------------------------------------------------
+filter {
+  dissect {
+    mapping => {
+      "message" => "%{ts} %{+ts} %{+ts} %{src} %{} %{prog}[%{pid}]: %{msg}"
```

Comment:
Let's use this example (which avoids the possibly confusing empty  `%{}` usage):

```
mapping => { "message" => "%{ts} %{+ts} %{+ts} %{src} %{prog}[%{pid}]: %{msg}" 
```

Sample input:

```
Apr 26 12:20:02 localhost systemd[1]: Starting system activity accounting tool...
```

Sample output (as through `stdout { codec => rubydebug }`:

```
{
  "msg"        => "Starting system activity accounting tool...",
  "@timestamp" => 2017-04-26T19:33:39.257Z,
  "src"        => "localhost",
  "@version"   => "1",
  "host"       => "localhost.localdomain",
  "pid"        => "1",
  "message"    => "Apr 26 12:20:02 localhost systemd[1]: Starting system activity accounting tool...",
  "type"       => "stdin",
  "prog"       => "systemd",
  "ts"         => "Apr 26 12:20:02"
}
```